### ISSUE TITLE
Build fails Tor framework from Carthage

### ISSUE THREAD

**Issue by ali-ht10** [iCepa/iCepa#4]
In the cartfile, I wrote **github "iCepa/Tor.framework" "master"** and ran **carthage update**.
The build fails with the following error:
*** Building scheme "Tor-Mac" in Tor.xcodeproj
** BUILD FAILED **

The following build commands failed:
	ExternalBuildToolExecution openssl-Mac
(1 failure)
A shell task failed with exit code 65:
** BUILD FAILED **

The following build commands failed:
	ExternalBuildToolExecution openssl-Mac
(1 failure)

I don't know what I'm doing wrong.

Please help. Thanks

**Comment by ali-ht10** [comment]
Update:

The following is from the build log,

cp: /Xcode/DerivedData/Tor-aqvflbpdbbukfahhupszphsxowds/Build/Intermediates/Tor.build/Release/openssl-x86_64/lib/*.a: No such file or directory

fatal error: /Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: can't open input file: /Xcode/DerivedData/Tor-aqvflbpdbbukfahhupszphsxowds/Build/Products/Release/libssl.*.a (No such file or directory)

fatal error: /Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: can't open input file: /Xcode/DerivedData/Tor-aqvflbpdbbukfahhupszphsxowds/Build/Products/Release/libcrypto.*.a (No such file or directory)

rm: /Xcode/DerivedData/Tor-aqvflbpdbbukfahhupszphsxowds/Build/Products/Release/libssl.*.a: No such file or directory

rm: /Xcode/DerivedData/Tor-aqvflbpdbbukfahhupszphsxowds/Build/Products/Release/libcrypto.*.a: No such file or directory

**Comment by conradev** [comment]
Can you post the full build log somewhere? The above is not particularly useful for debugging the issue.

Do you have `autoconf` and `automake` installed?

**Comment by mtigas** [comment]
Might be due to a known issue with the new XCode 7.3 and Tor 0.2.7.X. See https://github.com/ursachec/CPAProxy/issues/40 and https://trac.torproject.org/projects/tor/ticket/18626 There's a patch in that Tor ticket (and it’s been fixed in 0.2.8.2-alpha, though we might not get a fix for 0.2.7.x).

Though it could always be something else; could definitely use the full build output from @ali-ht10 to diagnose.

**Comment by ali-ht10** [comment]
@mtigas  I have XCode 7.3.
@conradev autoconf and automake is installed.

PFA the log file

[carthage-xcodebuild.rtf.zip](https://github.com/iCepa/iCepa/files/212284/carthage-xcodebuild.rtf.zip)

**Comment by ali-ht10** [comment]
Thanks for the reply.

@conradev I have automake and autoconf installed.
@mtigas I have XCode 7.3

PFA the log file,

[carthage-xcodebuild.rtf.zip](https://github.com/iCepa/iCepa/files/212294/carthage-xcodebuild.rtf.zip)

**Comment by conradev** [comment]
It seems to compile fine, just runs into an error installing:

```
/bin/sh: /Users/Downloads/iCepa-master: is a directory
make: *** [install_docs] Error 126
```

Will attempt to reproduce this!

**Comment by ali-ht10** [comment]
Thanks alot @conradev

**Comment by ali-ht10** [comment]
I can't seem to understand how tun2tor is working in this project. Can you give me any directions so that it can help me understand how tun2tor is working in this project.

Thanks

**Comment by ali-ht10** [comment]
Apparently the openssl issue was fixed because I had a space in the directory path. Once I removed that openssl compiled and installed successfully.

But libevent is causing an issue. This is the error:
configure.ac:101: error: possibly undefined macro: AC_PROG_LIBTOOL
configure: error: cannot find install-sh, install.sh, or shtool in "." "./.." "./../.."

I did ran **brew install automake autoconf**. automake and autoconf are in /usr/local/bin/. But I can't seem to understand the README requirements section in Tor.framework, **autoconf and automake in your PATH**. Where should autoconf and automake be installed? What PATH should I need to specify?

PFA the updated log file

[carthage-xcodebuild-update.rtf.zip](https://github.com/iCepa/iCepa/files/214848/carthage-xcodebuild-update.rtf.zip)

Thanks again

**Comment by conradev** [comment]
Thanks for discovering that! I will need to make sure that the build scripts function with spaces in the path.

If you run `echo $PATH` in Terminal and the directory that contains `automake` or `autoconf` is printed, then they are in your PATH. According to the script, your copies of `automake` and `autoconf` are in your PATH correctly.

It seems that I missed a required dependency! You also need `libtool` in your PATH. Run `brew install libtool` to fix that.

**Comment by conradev** [comment]
What questions do you have about `tun2tor` specifically?

**Comment by ali-ht10** [comment]
Thanks for the response.

Running **brew install libtool** did however worked in a way.** "Tor-Mac"** scheme built successfully. But there were errors when **"Tor-iOS"** scheme was building.

Here is the error which showed up on terminal:
*** Building scheme "Tor-Mac" in Tor.xcodeproj
*** Building scheme "Tor-iOS" in Tor.xcodeproj
** BUILD FAILED **


The following build commands failed:
	Ld /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Intermediates/Tor.build/Release-iphoneos/Tor-iOS.build/Objects-normal/arm64/Tor normal arm64
(1 failure)
configure.ac:14: warning: AM_INIT_AUTOMAKE: two- and three-arguments forms are deprecated.  For more info, see:
regress_buffer.c:199:12: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
regress.c:715:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress.c:716:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress.c:717:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_bufferevent.c:772:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_bufferevent.c:773:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_thread.c:379:8: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_thread.c:379:21: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_thread.c:393:3: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
/Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/shlib_wrap.sh: line 96: /Users/Downloads/iCepa-master/Carthage/Checkouts/Tor.framework/Tor/openssl/util/../apps/openssl: Undefined error: 0
regress_buffer.c:199:12: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
regress_buffer.c:199:12: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
regress.c:715:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress.c:716:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress.c:717:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_bufferevent.c:772:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_bufferevent.c:773:2: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_thread.c:379:8: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_thread.c:379:21: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
regress_thread.c:393:3: warning: absolute value function 'abs' given an argument of type 'long' but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
configure: error: Missing libraries; unable to proceed.
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor-trunnel.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor-trunnel.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libcurve25519_donna.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libcurve25519_donna.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libtor.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libtor.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor-event.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor-event.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libed25519_ref10.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libed25519_ref10.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libed25519_donna.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libed25519_donna.a (1 slices)
ld: warning: ignoring file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor-crypto.a, missing required architecture arm64 in file /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Products/Release-iphoneos/libor-crypto.a (1 slices)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
A shell task failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
	Ld /Users/Library/Developer/Xcode/DerivedData/Tor-cqwwpuougygethcwxgmrvujovwzm/Build/Intermediates/Tor.build/Release-iphoneos/Tor-iOS.build/Objects-normal/arm64/Tor normal arm64
(1 failure)

Questions for tun2tor:

- How does tun2tor parse packets?

- How can I use tun2tor to send the traffic to tor?

Link to log file: https://gist.github.com/ali-ht10/b9ad6eec872c9928a9c14c9907ce5328

Thanks again :D

**Comment by ali-ht10** [comment]
Looking at the log file, I noticed this:
"checking for openssl directory... configure: WARNING: Could not find a linkable openssl.  If you have it installed somewhere unusual, you can specify an explicit path using --with-openssl-dir
configure: error: Missing libraries; unable to proceed.
make: *** No targets specified and no makefile found.  Stop."

I can't figure out why openssl directory is missing.

**Comment by ali-ht10** [comment]
Basically what I want to achieve is that all my internet traffic on the device routes through Tor. Is it even possible to do that? Does this involve working at the network layer?

Thanks

**Comment by conradev** [comment]
I went ahead and fixed all of the issues you were seeing in the build:
- I added `libtool` and `gettext` to the `README`
- I made the build work in directories with spaces
- I fixed the `Could not find a linkable openssl` issue (backported from [here](https://gitweb.torproject.org/tor.git/commit/?id=049445bca33e8578757639fc8a236540fd0fd154) as @mtigas suggested)

Let me know if it builds successfully for you now!

I would like to point out the following from the `README`:

> The project does *not* work yet, and is in progress.

The goal of the project is indeed to route all internet traffic through Tor, but because it is not finished yet it does not yet achieve that goal. 

> How does tun2tor parse packets?

I wrote simple IPv4, IPv6, UDP and TCP packet parsers as a part of `tun2tor`.

> How can I use tun2tor to send the traffic to tor?

`tun2tor` isn't fully functional yet, but it has a SOCKS client that it will use to send traffic to `tor`. Right now, `tun2tor` only supports DNS traffic.
